### PR TITLE
Implement Cloud-Init NoCloud datasource

### PR DESCRIPTION
**Pull request by ivan4th**
- [X] basic support for passing `meta-data` and `user-data` from flexvolume definition
- [x] get rid of getvolumename
- [x] move volume population code to virtlet so the nodes don't need `genisoimage`
- [x] fix flexvolume unmount (kubelet fails to detect flexvolume mountpoint, deleted pods stuck in `Terminating`)
- [x] add integration test
- [x] update e2e test
- [x] update documentation

---
This change is [Reviewable](https://reviewable.io/reviews/mirantis/virtlet/264)

### CODE REVIEW

File: examples/ubuntu-vm.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      requiredDuringSchedulingIgnoredDuringExecution:
+        nodeSelectorTerms:
+        - matchExpressions:
+          - key: extraRuntime
+            operator: In
+            values:
+            - virtlet
+  containers:
+    - name: ubuntu-vm
+      # image: virtlet/github.com/eprasad/virt-cirros/raw/master/virt-cirros-0.3.4-x86_64-disk.img
```
Comment (jellonek) [inline examples/ubuntu-vm.yaml]: These comments can be probably removed for `ubuntu-vm.yaml`.

File: build/cmd.sh [hunk 2]
```diff
@@ -123,6 +123,19 @@ function clean {
     done
 }
 
+function gotest {
+    start_libvirt=""
+    subdir="$(virtlet_subdir)"
+    if [[ ${subdir} =~ /integration$ ]]; then
+        start_libvirt="VIRTLET_DISABLE_KVM=1 /start.sh -novirtlet && "
```
Comment (jellonek) [inline build/cmd.sh]: Can this VIRTLET_DISABLE_KVM be gathered from env? It hurts me a bit ;)

File: examples/ubuntu-vm.yaml (hunk 1 above)
Comment (ivan4th) [inline examples/ubuntu-vm.yaml]: Done.

File: build/cmd.sh (hunk 2 above)
Comment (ivan4th) [inline build/cmd.sh]: Will fix this

**Comment by jellonek** [comment]
Reviewed 14 of 21 files at r1, 4 of 6 files at r3, 21 of 21 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[Dockerfile.build, line 18 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8z1_q3pedA5laoYDl:-Kj8z1_q3pedA5laoYDm:b3qbmkn) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/Dockerfile.build#L18)):*
> ```
>                 libguestfs-tools \
>                 genisoimage \
>                 p7zip-full \
> ```

Is this package still needed? I can not find where it's used... (genisoimage is used in integration tests, so this is clear for me).

---

*[pkg/flexvolume/flexvolume.go, line 51 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8vdnfX0lxxqQlCyAw:-Kj8vdnfX0lxxqQlCyAx:b-mkknc0) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/pkg/flexvolume/flexvolume.go#L51)):*
> ```Go
> 	// XXX: invent a better wat to decide whether debugging should
> 	// be used for flexvolume driver. For now we only enable it if
> 	// Docker-in-Docker env is used
> ```

Imo this should be documented fixed location declared as constant value. Can be passed to pod as volume, to deliver info about wanted debugging logs.

---

*[pkg/utils/files.go, line 44 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8xHjbA-Jakymy1fgB:-Kj8xHjbA-Jakymy1fgC:bvifcrr) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/pkg/utils/files.go#L44)):*
> ```Go
> }
> 
> func GenIsoImage(isoPath string, volumeId string, srcDir string) error {
> ```

Exported func without docstring (i know that we have a lot of them, but would be nice to not introduce new).

---

*[pkg/utils/json.go, line 2 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8xUycxIE3B2g8Ge4B:-Kj8xUycxIE3B2g8Ge4C:bb6ht2h) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/pkg/utils/json.go#L2)):*
> ```Go
> /*
> Copyright 2016 Mirantis
> ```

2017

---

*[pkg/utils/json.go, line 24 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8xWS9kh8bzZCqdKP0:-Kj8xWS9kh8bzZCqdKP1:bcnghd0) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/pkg/utils/json.go#L24)):*
> ```Go
> )
> 
> func MapToJson(m map[string]interface{}) string {
> ```

Missing docstring.

---

*[tests/e2e/e2e.sh, line 29 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8xepwDYDoKjSMNgh9:-Kj8xepwDYDoKjSMNghA:b-59yvxb) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/tests/e2e/e2e.sh#L29)):*
> ```Shell
> SCRIPT_DIR="$(cd $(dirname "$(readlinkf "${BASH_SOURCE}")"); pwd)"
> virsh="${SCRIPT_DIR}/../../examples/virsh.sh"
> vmssh="${SCRIPT_DIR}/../../examples/vmssh.sh"
> ```

Looks like we should have a follow up which will move these scripts into other directory... `utils`, `contrib`, `scripts`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/264)*

**Comment by ivan4th** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[Dockerfile.build, line 18 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8z1_q3pedA5laoYDl:-Kj9-lE4Kc0ktt08kCk-:bccb4tw) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/Dockerfile.build#L18)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Is this package still needed? I can not find where it's used... (genisoimage is used in integration tests, so this is clear for me).
</blockquote></details>

It's used by pkg/utils/testing/files.go to unpack an iso image for testing purposes w/o loop mounting it. It could be documented but I think in this case we must document all the packages used here

---

*[pkg/flexvolume/flexvolume.go, line 51 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8vdnfX0lxxqQlCyAw:-Kj907PDcZtH_ll5DRIA:b-o0woo7) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/pkg/flexvolume/flexvolume.go#L51)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Imo this should be documented fixed location declared as constant value. Can be passed to pod as volume, to deliver info about wanted debugging logs.
</blockquote></details>

Well I actually think we may just try and invent some better way of turning on the debugging here, this one's a hack (although quite handy one). E.g. convert virtlet daemonsets etc. to templatable helm and make an option to enable the flexvolume debugging (that will create a magic file under the hood). I'm not sure documenting this hack is a good thing...

---

*[pkg/utils/files.go, line 44 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8xHjbA-Jakymy1fgB:-Kj90hdfrYDBWuh-yXPs:bcpgfmx) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/pkg/utils/files.go#L44)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Exported func without docstring (i know that we have a lot of them, but would be nice to not introduce new).
</blockquote></details>

Will fix, thanks

---

*[pkg/utils/json.go, line 2 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8xUycxIE3B2g8Ge4B:-Kj90j4K1Diurle17Srd:b-bqivcu) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/pkg/utils/json.go#L2)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

2017
</blockquote></details>

Will fix.

---

*[pkg/utils/json.go, line 24 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8xWS9kh8bzZCqdKP0:-Kj90lcSbLVQamUpzd1g:b-bqivcu) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/pkg/utils/json.go#L24)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Missing docstring.
</blockquote></details>

Will fix.

---

*[tests/e2e/e2e.sh, line 29 at r4](https://reviewable.io:443/reviews/mirantis/virtlet/264#-Kj8xepwDYDoKjSMNgh9:-Kj90nl2F9lmG0nrsGMd:b9qg0qv) ([raw file](https://github.com/mirantis/virtlet/blob/dc28be7b0bd1e528ba8f915cfebd8c83d53136b0/tests/e2e/e2e.sh#L29)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Looks like we should have a follow up which will move these scripts into other directory... `utils`, `contrib`, `scripts`?
</blockquote></details>

At the moment they're tied to examples one way or another (e.g. vmssh.sh uses predefined ssh key by default). I think we may do this cleanup in a followup PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/264)*

**Comment by jellonek** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/264#-:-KjCl-8_-odJJVnhSP5W:bnfp4nl)*

**Comment by vefimova** [comment]
Reviewed 13 of 21 files at r1, 3 of 6 files at r3, 20 of 21 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[build/cmd.sh, line 132 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/264#-KjCqehJeyF30Q8M8vH6:-KjCqehKpZorUVoonNsb:b-dq71os) ([raw file](https://github.com/mirantis/virtlet/blob/f9931a4f6c0a9891be26e042b4945dd6c53443bf/build/cmd.sh#L132)):*
> ```Shell
>       start_libvirt="VIRTLET_DISABLE_KVM=${VIRTLET_DISABLE_KVM:-} /start.sh -novirtlet && "
>     fi
>     vcmd "${start_libvirt}cd '${subdir}' && CGO_CFLAGS='-g -O2 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include' CGO_LDFLAGS='-lglib-2.0 -lvirt' go test $*"
> ```

CGO flags now can be omitted.

---

*[build/cmd.sh, line 136 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/264#-KjCqlcC9u3NiPUxVsaD:-KjCqlcC9u3NiPUxVsaE:bribihm) ([raw file](https://github.com/mirantis/virtlet/blob/f9931a4f6c0a9891be26e042b4945dd6c53443bf/build/cmd.sh#L136)):*
> ```Shell
> 
> function gobuild {
>     vcmd "cd '$(virtlet_subdir)' && CGO_CFLAGS='-g -O2 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include' CGO_LDFLAGS='-lglib-2.0 -lvirt' go build $*"
> ```

As above.

---

*[docs/volumes.md, line 143 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/264#-KjCvzgQRq60eUu7b2lv:-KjCvzgQRq60eUu7b2lw:b-whfrs2) ([raw file](https://github.com/mirantis/virtlet/blob/f9931a4f6c0a9891be26e042b4945dd6c53443bf/docs/volumes.md#L143)):*
> ```Markdown
> Here we set hostname for the VM, inject an ssh public key and provide a script that executed by cloud-init. There's helper script named [examples/vmssh.sh](../examples/vmssh.sh) that can be used to access VMs over ssh after providing ssh keys for them (it defaults to using a sample key from examples/ directory).
> 
> Virtlet's cloud-init mechanism is not finalized yet and will change in future so as to support dynamic metadata generation and passing metadata via the metadata server instead of `NoCloud`.
> ```

Why instead? I thought "along with". Aren't we going to keep 'nocloud' still in place after adding support for metadata server as well?

---

*[examples/README.md, line 38 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/264#-KjD-2924GTW1zyiF_8L:-KjD-2924GTW1zyiF_8M:bvrwdbz) ([raw file](https://github.com/mirantis/virtlet/blob/f9931a4f6c0a9891be26e042b4945dd6c53443bf/examples/README.md#L38)):*
> ```Markdown
> Besides [cirros-vm.yaml](cirros-vm.yaml), there's also [ubuntu-vm.yaml](ubuntu-vm.yaml) that can be used to start an Ubuntu Xenial VM. It can also be accessed using `vmssh.sh` after it boots:
> ```
> ./vmssh.sh root@ubuntu-vm
> ```

missed [command...]

---

*[tests/integration/container_test.go, line 183 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/264#-KjD2-ihtOZBpj7zot8j:-KjD2-ihtOZBpj7zot8k:b-qqs7hx) ([raw file](https://github.com/mirantis/virtlet/blob/f9931a4f6c0a9891be26e042b4945dd6c53443bf/tests/integration/container_test.go#L183)):*
> ```Go
> func (ct *containerTester) verifyNoContainers(filter *kubeapi.ContainerFilter) {
> 	if len(ct.listContainers(filter).Containers) != 0 {
> 		ct.t.Errorf("expected no containers to be listed")
> ```

I suppose, the error msg should contain info about set filter set for List request.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/264)*

**Comment by ivan4th** [comment]
Review status: 33 of 37 files reviewed at latest revision, 5 unresolved discussions.

---

*[build/cmd.sh, line 132 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/264#-KjCqehJeyF30Q8M8vH6:-KjDKLgNNvJ630HuoFAq:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/f9931a4f6c0a9891be26e042b4945dd6c53443bf/build/cmd.sh#L132)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

CGO flags now can be omitted.
</blockquote></details>

Done.

---

*[build/cmd.sh, line 136 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/264#-KjCqlcC9u3NiPUxVsaD:-KjDKM1faFoIdzc9s1I3:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/f9931a4f6c0a9891be26e042b4945dd6c53443bf/build/cmd.sh#L136)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

As above.
</blockquote></details>

Done.

---

*[examples/README.md, line 38 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/264#-KjD-2924GTW1zyiF_8L:-KjDKOnHtrus2K0SUL7I:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/f9931a4f6c0a9891be26e042b4945dd6c53443bf/examples/README.md#L38)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

missed [command...]
</blockquote></details>

Done.

---

*[tests/integration/container_test.go, line 183 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/264#-KjD2-ihtOZBpj7zot8j:-KjDKNrDAxcGHXvJZNKD:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/f9931a4f6c0a9891be26e042b4945dd6c53443bf/tests/integration/container_test.go#L183)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

I suppose, the error msg should contain info about set filter set for List request.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/264)*

**Comment by ivan4th** [comment]
Review status: 33 of 37 files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/volumes.md, line 143 at r5](https://reviewable.io:443/reviews/mirantis/virtlet/264#-KjCvzgQRq60eUu7b2lv:-KjDKpFqAyP2d56SvPEs:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/f9931a4f6c0a9891be26e042b4945dd6c53443bf/docs/volumes.md#L143)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

Why instead? I thought "along with". Aren't we going to keep 'nocloud' still in place after adding support for metadata server as well?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/264)*

**Comment by vefimova** [comment]
:lgtm:

---

Review status: 33 of 37 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/264#-:-KjDLWLc9sCWcAymDo8o:bnfp4nl)*

**Comment by ivan4th** [comment]
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/264)*

**Pull request merged**